newm-server/src/main/kotlin/io/newm/server/features/release/repo/OutletReleaseRepositoryImpl.kt

```diff
@@ -0,0 +1,100 @@
+package io.newm.server.features.release.repo
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import io.ktor.client.HttpClient
+import io.ktor.client.plugins.retry
+import io.ktor.client.request.accept
+import io.ktor.client.request.basicAuth
+import io.ktor.client.request.bearerAuth
+import io.ktor.client.request.forms.submitForm
+import io.ktor.client.request.get
+import io.ktor.http.ContentType
+import io.ktor.http.append
+import io.ktor.http.parameters
+import io.ktor.http.parametersOf
+import io.ktor.server.application.ApplicationEnvironment
+import io.newm.server.features.release.model.SpotifyAuthResponse
+import io.newm.server.features.release.model.SpotifySearchResponse
+import io.newm.server.features.song.repo.SongRepository
+import io.newm.server.ktx.checkedBody
+import io.newm.server.ktx.getSecureConfigString
+import io.newm.shared.koin.inject
+import io.newm.shared.ktx.debug
+import io.newm.shared.ktx.getConfigString
+import io.newm.shared.ktx.info
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import org.koin.core.parameter.parametersOf
+import org.slf4j.Logger
+import java.time.Duration
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+private const val SPOTIFY_SEARCH_API_URL = "https://api.spotify.com/v1/search"
+private const val EXPIRY_WINDOW_MILLIS = 300_000L
+private const val API_TOKEN_CACHE_KEY = -1
+
+private data class ApiTokenCacheEntry(val token: String, val expiryTimeMillis: Long)
+
+internal class OutletReleaseRepositoryImpl(
+    private val environment: ApplicationEnvironment,
+    private val httpClient: HttpClient,
+    private val songRepository: SongRepository
+
+) : OutletReleaseRepository {
+    private val logger: Logger by inject { parametersOf(javaClass.simpleName) }
+    private val apiTokenMutex = Mutex()
+    private val apiTokenCache = Caffeine.newBuilder()
+        .expireAfterWrite(Duration.ofHours(1L))
+        .build<Int, ApiTokenCacheEntry>()
+
+    override suspend fun isSongReleased(songId: UUID): Boolean {
+        val isrc = songRepository.get(songId).isrc!!.lowercase().replace("-", "")
+        logger.debug { "isSongReleased: songId = $songId, isrc = $isrc" }
+
+        val response = httpClient.get(SPOTIFY_SEARCH_API_URL) {
+            url {
+                with(parameters) {
+                    append("market", "US")
+                    append("limit", "1")
+                    append("type", "track")
+                    append("q", "isrc:$isrc")
+                }
+            }
+            bearerAuth(getApiToken())
+            accept(ContentType.Application.Json)
+        }.checkedBody<SpotifySearchResponse>()
+        return response.tracks.total > 0
+    }
+
+    private suspend fun getApiToken(): String = apiTokenMutex.withLock {
```

Comment:
See if there's a way to share the authentication between this repository and the SpotifyProfileUrlVerifier.

Reply:
I thought about doing that and desisted. I'll get it done.